greymatter-io/operator
# Multiple operators in one cluster

[sc-20738]

Companion PR in gitops-core: https://github.com/greymatter-io/gitops-core/pull/72
(Most of the work is there)

## How to test

0. Checkout this branch locally.
1. Build the operator and push your own tag to quay. The examples in this instructions assume your tag is named `multi-operator`.
2. Point the submodule at the companion PR's branch. This should do it: `git submodule update --init --recursive`
3. Make your first operator deployment as usual. This should do it:
```
(
cd pkg/cuemodule/core

kubectl create namespace gm-operator

kubectl create secret docker-registry gm-docker-secret \
  --docker-server=quay.io \
  --docker-username=$QUAY_USERNAME \
  --docker-password=$QUAY_PASSWORD \
  -n gm-operator

kubectl create secret generic greymatter-sync-secret \
  --from-file=id_ed25519=$HOME/.ssh/id_ed25519 \
  -n gm-operator

cue eval -c ./k8s/outputs --out text \
         -t operator_image=quay.io/greymatterio/operator:multi-operator \
         -e operator_manifests_yaml | kubectl apply -f -
)
```
4. Confirm the operator is up and running, and that a mesh gets deployed as before. Since several K8s config objects have been adjusted to have `operator_namespace`-scoped names, this is a test in itself.
5. Now branch the submodule for the second operator deployment, and make the changes discussed in this comment: https://app.shortcut.com/grey-matter/story/20738/support-for-deploying-multiple-operators-into-one-cluster-so-that-you-get-2-meshes-in-a-single-cluster#activity-24742
6. Don't forget to change the operator's StatefulSet to point at your new branch by editing  `pkg/cuemodule/core/outputs/k8s/operator.cue`, and push your new branch up to gitops-core, so that the second operator will fetch from the second branch. (I have made the mistake of leaving this pointed at the old branch a couple times, and it results in log messages about a failure to get leader-election rights in the first namespace, because of course the operator believes it should be installed wherever the pushed GitOps repository says it should be.)
7. Deploy your second operator. This should do it, and assumes your `config.operator_namespace` is `second-gm-operator`. (Notice this command isn't actually much different otherwise, because the changes that make it different are in the CUE (of the submodule). It was only necessary to change the namespace for the manual secret creations):
```
(
cd pkg/cuemodule/core

kubectl create namespace second-gm-operator

kubectl create secret docker-registry gm-docker-secret \
  --docker-server=quay.io \
  --docker-username=$QUAY_USERNAME \
  --docker-password=$QUAY_PASSWORD \
  -n second-gm-operator

kubectl create secret generic greymatter-sync-secret \
  --from-file=id_ed25519=$HOME/.ssh/id_ed25519 \
  -n second-gm-operator

cue eval -c ./k8s/outputs --out text \
         -t operator_image=quay.io/greymatterio/operator:multi-operator \
         -e operator_manifests_yaml | kubectl apply -f -
)
```
8. Confirm the second operator deploys into its own namespace (`second-gm-operator`), and deploys a mesh into the namespace you put in inputs.cue. You should be able to get to the second dashboard with a second port-forward into the second edge. 🥈 
9. Confirm that something deployed into a given watched_namespace is only noticed by the operator watching that namespace, and that both operators are watching their respective namespaces. (Deploy two workloads, one into a namespace watched by operator1, and another into a namespace watched by operator2.) Then delete these workloads, and confirm they are deleted by the correct operator. You can watch the logs and confirm the workloads eventually end up with a box on the dashboard which is cleaned up after deletion.
10. Tear down one mesh, and confirm the second mesh is not destroyed. You can do this with something like
```
(
cd pkg/cuemodule/core
cue eval -c ./k8s/outputs --out text -e operator_manifests_yaml | kubectl delete -f -
)
```
11. Tear down the other mesh and confirm that everything is cleaned up as before this PR. (Everything but your manually created namespaces and secrets)
12. Test a multi-operator deployment with Spire enabled by carefully redoing steps 2-11.
